blog/tests.py

```diff
@@ -1,3 +1,183 @@
+    response = self.client.post(f"/post/{published_post.pk}/edit/", {"title": "new title", "text": "new text"}) 
+    self.assertRedirects(response, f"/post/{published_post.pk}/")
+    published_post.refresh_from_db()
+    self.assertEqual(published_post.title, "new title")
+    self.assertEqual(published_post.text, "new text")
+
+  def test_post_unpublished(self):
+    """If a post is unpublished then the user should not be able to edit it"""
+    author = User.objects.create(username = "user1")
+    unpublished_post = Post.objects.create(title = "Post02", text = "Great", author = author)
+    response = self.client.post(f"/post/{unpublished_post.pk}/edit/", {"title": "new title", "text": "new text"})
+    self.assertRedirects(response, f"/post/{unpublished_post.pk}/")
+    unpublished_post.refresh_from_db()
+    self.assertEqual(unpublished_post.title, "new title")
+    self.assertEqual(unpublished_post.text, "new text")
+
+  def test_post_loggedout(self):
+    """If user is logged they won't be able to edit posts"""
+    self.client.logout()
+    author = User.objects.create(username = "user1") 
+    published_post = Post.objects.create(title = "Post01", text = "Hey", author = author, published_date = now())
+    response = self.client.post(f"/post/{published_post.pk}/edit/") 
+    self.assertEqual(response.status_code, 404) 
+    unpublished_post = Post.objects.create(title = "Post02", text = "Great", author = author)
+    response = self.client.post(f"/post/{unpublished_post.pk}/edit/") 
+    self.assertEqual(response.status_code, 404) 
+
+class PostNewTest (TestCase):
+  def setUp(self): # setting up username and password for every test
+      user = User.objects.create(username = "test_user")
+      user.set_password("password123")
+      user.save()
+      self.client.login(username = "test_user", password = "password123")
+
+  def test_get(self):
+    """Getting the page to create a new post"""
+    response = self.client.get('/post/new/') 
+    self.assertEqual(response.status_code, 200)
+
+  def test_post_invalid_data(self):
+    """Getting the page to create a new post"""
+    response = self.client.post('/post/new/')
+    self.assertEqual(response.status_code, 200)
+    self.assertEqual(response.context["form"].errors, {"title": ["This field is required."], "text": ["This field is required."]})
+
+  def test_post_valid_data(self):
+    response = self.client.post('/post/new/', {"title": "new title", "text": "new text"})
+    new_post = Post.objects.get(title = "new title", text = "new text")
+    self.assertRedirects(response,f"/post/{new_post.pk}/")
+
+  def test_post_loggedout(self):
+    """If user is not logged in they should not create new posts"""
+    self.client.logout() 
+    response = self.client.post("/post/new/")
+    self.assertEqual(response.status_code, 404)
+    response = self.client.post("/post/new/", {"title": "new title", "text": "new text"})
+    self.assertEqual(response.status_code, 404)
+
```

Comment:
After this test, could you add a test that checks what happens if someone tries to create a post with the same title as another post that already exists?